docs/website/docs/v1.x/dsl/middleware.md

```diff
@@ -3,4 +3,422 @@ sidebar_position: "8"
+
+Or in short, aspect is a common concern required throughout the application, and its implementation could lead to repeated boilerplate code and in violation of the principle of separation of concerns.
+There is a paradigm in the programming world called [aspect-oriented programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming) that aims for modular handling of these common concerns in an application. 
+
+Some examples of common "aspects" required throughout the application
+- logging,
+- timeouts (preventing long-running code)
+- retries (or handling flakiness for example while accessing third party APIs)
+- authenticating a user before using the REST resource (basic, or custom ones like OAuth / single sign-on, etc).
+
+This is where middleware comes to the rescue. 
+Using middlewares we can compose out-of-the-box middlewares (or our custom middlewares) to address the above-mentioned concerns using ++ and @@ operators as shown below.
+
+#### Cleaned up code using middleware to address cross-cutting concerns like auth, request/response logging, etc.
+```scala
+// compose basic auth, request/response logging, timeouts middlewares
+val composedMiddlewares = Middleware.basicAuth("user","pw") ++ 
+        Middleware.debug ++ 
+        Middleware.timeout(5 seconds) 
+
+private val app = Http.collectZIO[Request] {
+  case Method.GET -> !! / "users" / id =>
+    // core business logic  
+    dbService.lookupUsersById(id).map(Response.json(_.json))
+  case Method.GET -> !! / "users"    =>
+    // core business logic  
+    dbService.paginatedUsers(pageNum).map(Response.json(_.json))
+} @@ composedMiddlewares // attach composedMiddlewares to the app using @@
+```
+Observe how we gained following benefits by using middlewares
+* **Readability**: de-cluttering business logic.
+* **Modularity**: we can manage aspects independently without making changes in 100 places. For example, 
+  * replacing the logging mechanism from logback to log4j2 will require a change in one place, the logging middleware).
+  * replacing the authentication mechanism from OAuth to single sign-on will require changing the auth middleware
+* **Testability**: we can test our aspects independently.
+
+## Middleware in zio-http
+
+A middleware helps in addressing common crosscutting concerns without duplicating boilerplate code.
+
+#### Revisiting HTTP 
+[`Http`](https://dream11.github.io/zio-http/docs/v1.x/dsl/http) is the most fundamental type for modelling Http applications
+
+```Http[-R, +E, -A, +B]``` is equivalent to ```(A) => ZIO[R, Option[E], B]``` where
+
+* `R` type of Environment 
+* `E` type of the Error when the function fails with Some[E]
+* `A` is the type of the function parameter
+* `B` type of the result when function succeeds 
+
+Middleware is simply a function that takes one Http as a parameter and returns another Http,
+
+```Http => Http```
+
+So, a middleware represents function transformation f1 => f2
```

Comment:
This is great, I think we should also showcase that you can call a middleware like this — 

```scala
mid(http)
```
Which emphasises on the fact that it's really a function :)

Then later slowly incorporate the `@@` syntax.